superset/views/tags.py

```diff
@@ -0,0 +1,197 @@
+        tagged_objects = []
+        for name in request.get_json(force=True):
+            if ':' in name:
+                type_name = name.split(':', 1)[0]
+                type_ = TagTypes[type_name]
+            else:
+                type_ = TagTypes.custom
+
+            tag = db.session.query(Tag).filter_by(name=name, type=type_).first()
+            if not tag:
+                tag = Tag(name=name, type=type_)
+
+            tagged_objects.append(
+                TaggedObject(
+                    object_id=object_id,
+                    object_type=object_type,
+                    tag=tag,
+                ),
+            )
+
+        db.session.add_all(tagged_objects)
+        db.session.commit()
+
+        return Response(status=201)  # 201 CREATED
+
+    @expose('/tags/<object_type:object_type>/<int:object_id>/', methods=['DELETE'])
+    def delete(self, object_type, object_id):
+        """Remove tags from an object."""
+        tag_names = request.get_json(force=True)
+        if not tag_names:
+            return Response(status=403)
+
+        db.session.query(TaggedObject).filter(and_(
+            TaggedObject.object_type == object_type,
+            TaggedObject.object_id == object_id),
+            TaggedObject.tag.has(Tag.name.in_(tag_names)),
+        ).delete(synchronize_session=False)
+        db.session.commit()
+
+        return Response(status=204)  # 204 NO CONTENT
+
+    @expose('/tagged_objects/', methods=['GET', 'POST'])
+    def tagged_objects(self):
+        query = db.session.query(
+            TaggedObject,
+            superset.models.core.Dashboard,
+            superset.models.core.Slice,
+            SavedQuery,
+        ).join(Tag)
+
+        tags = request.args.get('tags')
+        if not tags:
+            return Response(
+                json.dumps([]), status=200, content_type='application/json')
+
+        tags = [process_template(tag) for tag in tags.split(',')]
+        query = query.filter(Tag.name.in_(tags))
+
+        # filter types
+        types = request.args.get('types')
+        if types:
+            query = query.filter(TaggedObject.object_type.in_(types.split(',')))
+
+        # get names
+        query = query.outerjoin(
+            superset.models.core.Dashboard,
+            and_(
+                TaggedObject.object_id == superset.models.core.Dashboard.id,
+                TaggedObject.object_type == ObjectTypes.dashboard,
+            ),
+        ).outerjoin(
+            superset.models.core.Slice,
+            and_(
+                TaggedObject.object_id == superset.models.core.Slice.id,
+                TaggedObject.object_type == ObjectTypes.chart,
+            ),
+        ).outerjoin(
+            SavedQuery,
+            and_(
+                TaggedObject.object_id == SavedQuery.id,
+                TaggedObject.object_type == ObjectTypes.query,
+            ),
+        ).group_by(TaggedObject.object_id, TaggedObject.object_type)
```

Comment:
Trying to understand what kind of SQL will come out of that query object... Is the group_by needed?

Also you may be able to parameterize the ObjectTypes as in:
```python
tagged_models = [Slice, Dashboard, Query]
for tm in tagged_models:
    query = query.outerjoin(tm, {...})
```
It would make it easier to add more tagged models in the future but may make the code more opaque.

Reply:
The group by is needed because a user might search for tags "a" and "b", and if an object is tagged with both it would return duplicates.